test/integration_tests/test_reuse.py

```diff
@@ -0,0 +1,176 @@
+        )
+
+    def test_reuse_partial(self):
+        select_query = """SELECT id, label FROM DETRAC JOIN
+            LATERAL YoloV5(data) AS Obj(label, bbox, conf) WHERE id < 5;"""
+
+        execute_query_fetch_all(select_query)
+        select_query = """SELECT id, label FROM DETRAC JOIN
+            LATERAL YoloV5(data) AS Obj(label, bbox, conf) WHERE id < 15;"""
+
+        reuse_batch = execute_query_fetch_all(select_query)
+        self._verify_reuse_correctness(select_query, reuse_batch)
+
+    def test_reuse_in_with_multiple_occurences(self):
+        select_query = """SELECT id, label FROM DETRAC JOIN
+            LATERAL YoloV5(data) AS Obj(label, bbox, conf) WHERE id < 10;"""
+
+        execute_query_fetch_all(select_query)
+
+        # multiple occurences of the same function expression
+        select_query = """SELECT id, YoloV5(data).labels FROM DETRAC JOIN
+            LATERAL YoloV5(data) AS Obj(label, bbox, conf) WHERE id < 15;"""
+
+        reuse_batch = execute_query_fetch_all(select_query)
+        self._verify_reuse_correctness(select_query, reuse_batch)
+
+        # enable these test cases when we increase the support of caching
+        if False:
+            # different query format
+            select_query = (
+                """SELECT id, YoloV5(data).labels FROM DETRAC WHERE id < 25;"""
+            )
+            reuse_batch = execute_query_fetch_all(select_query)
+            self._verify_reuse_correctness(select_query, reuse_batch)
+
+            # different query format
+            select_query = """SELECT id, YoloV5(data).scores FROM DETRAC WHERE ['car'] <@ YoloV5(data).labels AND id < 20"""
+            reuse_batch = execute_query_fetch_all(select_query)
+            self._verify_reuse_correctness(select_query, reuse_batch)
+
+    def test_reuse_does_not_work_when_expression_in_where_clause(self):
+        # add with subsequent PR
+        select_query = """
+            SELECT id FROM DETRAC
+            WHERE ArrayCount(YoloV5(data).labels, 'car') > 3 AND id < 5;"""
+        plan = next(get_logical_query_plan(select_query).find_all(LogicalFilter))
+        yolo_expr = None
+        for expr in plan.predicate.find_all(FunctionExpression):
+            if expr.name == "YoloV5":
+                yolo_expr = expr
+
+        self.assertFalse(yolo_expr.has_cache())
+
+    def test_reuse_after_server_shutdown(self):
+        select_query = """SELECT id, label FROM DETRAC JOIN
+            LATERAL YoloV5(data) AS Obj(label, bbox, conf) WHERE id < 10;"""
+        execute_query_fetch_all(select_query)
+
+        # Stop and restart server
+        os.system("nohup eva_server --stop")
+        os.system("nohup eva_server --start &")
+
+        select_query = """SELECT id, label FROM DETRAC JOIN
+            LATERAL YoloV5(data) AS Obj(label, bbox, conf) WHERE id < 15;"""
+
+        reuse_batch = execute_query_fetch_all(select_query)
+        self._verify_reuse_correctness(select_query, reuse_batch)
+
```

Comment:
Do we want to also check the time speedup with reuse?

Reply:
Might lead to a flaky test.